### DISCUSSION TITLE
Work Graphs

### DISCUSSION THREAD

**Discussion by Usnul** [gpuweb/gpuweb#4926] [Ideas/Feedback]
Not sure if this was discussed yet or not. But I wanted to kick-off a discussion about inclusion of WorkGraph API into WebGPU

# What are work graphs?
Very briefly, a it's an API for creating GPU compute pipelines with dynamic resource allocation. For example, if you want to build a mesh on the GPU, you could currently use a storage buffer likes so:
```wgsl
struct Elements{
  end: atomic<u32>,
  elements: array< Triangle >
}
```

The problem is that you have to pre-allocate such a buffer for worst-case, as you don't know ahead of time how many triangles you'd be generating.

There are other benefits such as better scheduling semantics to improve utilization.

# References:

DirectX spec: https://github.com/microsoft/DirectX-Specs/blob/master/d3d/WorkGraphs.md
DirectX blog: https://devblogs.microsoft.com/directx/d3d12-work-graphs/

Here's a nice talk from DirectX guys:
https://www.youtube.com/watch?v=0S8qymwcHIc

AMD talk:
https://www.youtube.com/watch?v=QQP6-JF64DQ

# Why is this relevant for the web?
* Resources are more scarce in general, especially on mobile. This limits what we can do quite severely, this feature would help alleviate this issue.
* We have a fairly sequential programming model which makes efficient GPU utilization difficult. Work Graphs would help here as well, by side-stepping the issue more or less.
* Free up main thread a little. Less work to be done on the CPU, this is especially relevant for allocations, which can take a while in my experience and cause hick-ups.

**Comment by kainino0x** [comment]
A WebGPU feature would need to be abstracted over D3D12, Metal, and Vulkan. Do you know what work-graph-like APIs look like in Metal/Vulkan (if any)?

**Reply by Usnul** [reply]
Quick search shows vulkan:
https://gpuopen.com/gpu-work-graphs-in-vulkan/
and sample https://github.com/GPUOpen-LibrariesAndSDKs/Vulkan-Samples


for metal, not sure, this seems like it might be similar tech https://developer.apple.com/documentation/metalperformanceshaders
but intention seems quite different.

**Comment by zwalden** [comment]
Is there any chance of this being put in for review; or a place I can follow this?

Was hoping to play around and prototype https://gpuopen.com/download/Real-Time_GPU_Tree_Generation.pdf in the browser, but it's simply not viable for real time generation without work graphs (atleast, that I can tell!) -- was, again, hoping to just find a place to keep an eye on it, if that's not here.

**Comment by Frendysemon** [comment]
I think it would be a nice addition. Better GPU scheduling and dynamic allocation could really help, especially on mobile where resources are limited. Just hope it stays simple enough for developers to use.
